scripts/field_report.py

```diff
@@ -0,0 +1,262 @@
+        """Prepare for analysis of C source files."""
+        if options.clang_library_file:
+            clang.cindex.Config.set_library_file(options.clang_library_file)
+        self.parse_options = []
+        if options.target:
+            self.parse_options += ['-target', options.target]
+        for d in options.include:
+            self.parse_options.append('-I' + d)
+        for d in options.define:
+            self.parse_options.append('-D' + d)
+        self.index = clang.cindex.Index.create()
+        self.files = {} #type: Dict[str, TranslationUnit]
+        # fields[TYPE_OR_STRUCT_NAME][FIELD_NAME]
+        self.fields = {} #type: Dict[str, Dict[str, FieldInfo]]
+
+    def load(self, *filenames: str) -> None:
+        """Load the AST of the given source files."""
+        for filename in filenames:
+            self.files[filename] = self.index.parse(filename,
+                                                    self.parse_options)
+
+    def in_interesting_file(self, location: SourceLocation) -> bool:
+        """Whether the given location is in a file that should be analyzed."""
+        if not hasattr(location.file, 'name'):
+            # Some artificial nodes have associated no file name.
+            # Let's hope they're not important.
+            return False
+        return location.file.name in self.files
+
+    def read_field_definitions(self, filenames: List[str]) -> None:
+        """Parse structure field definitions in the given C source files (usually headers)."""
+        self.load(*filenames)
+        for filename in filenames:
+            for node in self.files[filename].cursor.walk_preorder():
+                if not self.in_interesting_file(node.location):
+                    continue
+                if node.kind == CursorKind.FIELD_DECL:
+                    type_name = node.lexical_parent.spelling
+                    # Skip unions
+                    if node.lexical_parent.kind != CursorKind.STRUCT_DECL:
+                        continue
+                    # type_name is the struct name. If there's no struct
+                    # name, see if there's a typedef name, to cope with
+                    # "typedef struct { ... } foo;"
+                    if not type_name:
+                        type_name = node.lexical_parent.type.spelling
+                    # Skip anonymous structs for now.
+                    if '(anonymous ' in type_name:
+                        continue
+                    self.fields.setdefault(type_name, collections.OrderedDict())
+                    self.fields[type_name][node.spelling] = FieldInfo(node)
+
+    @staticmethod
+    def get_underlying_type(typ: clang.cindex.Type) -> Optional[clang.cindex.Type]:
+        """Strip off one level of type indirection.
+
+        Return None if the type is as primitive as can be.
+        """
+        if hasattr(typ, 'get_canonical'):
+            lower = typ.get_canonical()
+            if lower != typ:
```

Comment:
if `lower == typ` is that meant to return None? I wonder if we could get away with a `try/except` approach.

Reply:
The following cases cover `lower == typ` with a return other than `None`. I don't know what exception you're thinking of catching here.

The data structures I'm traversing here are not documented. I arrived at this code by trial and error.